sdk/src/main/scala/io/horizen/account/state/McAddrOwnershipMsgProcessor.scala

```diff
@@ -0,0 +1,351 @@
+      log.debug(msgStr)
+      throw new ExecutionRevertedException(msgStr)
+    }
+
+    val ownershipList = getListOfMcAddrOwnerships(view)
+    McAddrOwnershipDataListEncoder.encode(ownershipList.asJava)
+
+  }
+
+  override def getListOfMcAddrOwnerships(view: BaseAccountStateView, scAddressOpt: Option[String] = None): Seq[McAddrOwnershipData] = {
+    var ownershipsList = Seq[McAddrOwnershipData]()
+    var nodeReference = view.getAccountStorage(contractAddress, LinkedListTipKey)
+
+    while (!linkedListNodeRefIsNull(nodeReference)) {
+      val (item: McAddrOwnershipData, prevNodeReference: Array[Byte]) = getOwnershipListItem(view, nodeReference)
+      scAddressOpt match {
+        case Some(scAddr) =>
+          if (scAddr == item.scAddress)
+            ownershipsList = item +: ownershipsList
+
+        case None =>
+          ownershipsList = item +: ownershipsList
+      }
+      nodeReference = prevNodeReference
+    }
+    ownershipsList
+  }
+
+  override def ownershipDataExist(view: BaseAccountStateView, ownershipId: Array[Byte]): Boolean = {
+    // do the RAW-strategy read even if the record is actually multi-line in stateDb. It will save some gas.
+    val data = view.getAccountStorage(contractAddress, ownershipId)
+    // getting a not existing key from state DB using RAW strategy
+    // gives an array of 32 bytes filled with 0, while using CHUNK strategy
+    // gives an empty array instead
+    !data.sameElements(NULL_HEX_STRING_32)  }
+
+  private def isMcAddrAlreadyAssociated(view: BaseAccountStateView, mcAddress: String): Option[String] = {
+    var nodeReference = view.getAccountStorage(contractAddress, LinkedListTipKey)
+    while (!linkedListNodeRefIsNull(nodeReference)) {
+      val (item: McAddrOwnershipData, prevNodeReference: Array[Byte]) = getOwnershipListItem(view, nodeReference)
+      if (item.mcTransparentAddress.equals(mcAddress))
+        return Some(item.scAddress)
+      nodeReference = prevNodeReference
+    }
+    None
+  }
+
+  @throws(classOf[ExecutionFailedException])
+  override def process(msg: Message, view: BaseAccountStateView, gas: GasPool, blockContext: BlockContext): Array[Byte] = {
+    val gasView = view.getGasTrackedView(gas)
+    getFunctionSignature(msg.getData) match {
+      case AddNewOwnershipCmd => doAddNewOwnershipCmd(msg, gasView)
+      case RemoveOwnershipCmd => doRemoveOwnershipCmd(msg, gasView)
+      case GetListOfAllOwnershipsCmd => doGetListOfAllOwnershipsCmd(msg, gasView)
+      case GetListOfOwnershipsCmd => doGetListOfOwnershipsCmd(msg, gasView)
+
+      case opCodeHex => throw new ExecutionRevertedException(s"op code not supported: $opCodeHex")
+    }
+  }
+
+}
+
+object McAddrOwnershipMsgProcessor {
+
+  val LinkedListTipKey: Array[Byte] = Blake2b256.hash("OwnershipTip")
+  val LinkedListNullValue: Array[Byte] = Blake2b256.hash("OwnershipNull")
+
+  val AddNewOwnershipCmd: String = getABIMethodId("sendKeysOwnership(bytes3,bytes32,bytes1,bytes32,bytes32)")
```

Comment:
This doesn't seem correct because in AddNewOwnershipCmdInputDecoder class the third param is defined as Bytes24 and not Bytes1.

Reply:
Fixed